akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala

```diff
@@ -0,0 +1,168 @@
+      case Nil ⇒
+        Future.failed(ResolveFailedException(s"Timed out resolving $name with nameservers: $nameServers"))
+      case head :: tail ⇒ resolve(name, requestType, head).recoverWith {
+        case t: Throwable ⇒
+          log.error(t, "Resolve failed. Trying next name server")
+          resolve(name, requestType, tail)
+      }
+    }
+  }
+
+  private def sendQuestion(resolver: ActorRef, message: Question): Future[Seq[DnsResourceRecord]] = {
+    val result = (resolver ? message).mapTo[Answer].map(_.rrs)
+    result.onFailure {
+      case _: Throwable ⇒ resolver ! DropRequest(message.id)
+    }
+    result
+  }
+
+  private def resolve(name: String, requestType: RequestType, resolver: ActorRef): Future[DnsProtocol.Resolved] = {
+    log.debug("Attempting to resolve {} with {}", name, resolver)
+    val caseFoldedName = Helpers.toRootLowerCase(name)
+    val recs: Future[Seq[DnsResourceRecord]] = requestType match {
+      case Ip(ipv4, ipv6) ⇒
+        val ipv4Recs = if (ipv4)
+          cache.get((name, Ipv4Type)) match {
+            case Some(r) ⇒
+              log.debug("Ipv4 cached {}", r)
+              Future.successful(r)
+            case None ⇒
+              sendQuestion(resolver, Question4(nextId(), caseFoldedName))
+          }
+        else
+          Empty
+
+        val ipv6Recs = if (ipv6)
+          cache.get((name, Ipv6Type)) match {
+            case Some(r) ⇒
+              log.debug("Ipv6 cached {}", r)
+              Future.successful(r)
+            case None ⇒
+              sendQuestion(resolver, Question6(nextId(), caseFoldedName))
+          }
+        else
+          Empty
+
+        ipv4Recs.flatMap(ipv4Records ⇒ {
+          // TODO, do we want config to specify a max for this?
+          if (ipv4Records.nonEmpty) {
+            val minTtl4 = ipv4Records.minBy(_.ttl).ttl
+            cache.put((name, Ipv4Type), ipv4Records, minTtl4)
+          }
+          ipv6Recs.map(ipv6Records ⇒ {
+            if (ipv6Records.nonEmpty) {
+              val minTtl6 = ipv6Records.minBy(_.ttl).ttl
+              cache.put((name, Ipv6Type), ipv6Records, minTtl6)
+            }
+            ipv4Records ++ ipv6Records
+          })
+        })
+      case Srv ⇒
+        cache.get((name, Ipv4Type)) match {
+          case Some(r) ⇒ Future.successful(r)
+          case None ⇒
+            sendQuestion(resolver, SrvQuestion(nextId, caseFoldedName))
+        }
+
+    }
+    // FIXME, do we need this? it was in the old version, no test covered it.
+    /*
+    val canonicalNames: Map[String, String] = all.collect({
+      case CNameRecord(incomingName, _, cName) =>
+        incomingName.toLowerCase -> cName.toLowerCase
+    })(breakOut)
+
+    @tailrec
+    def resolveCanonicalName(name: String): String = {
+      canonicalNames.get(name) match {
+        case Some(cName) => resolveCanonicalName(cName)
+        case None => name
+      }
+    }
+
+    val canonicalName = resolveCanonicalName(name)
+    */
```

Comment:
Huh? I searched the current codebase for `canonicalNames`/`canonicalName` but get no hits for it?

Reply:
https://github.com/ilya-epifanov/akka-dns/blob/adbaf294db6c08dadc6eef061860e2329a43b62a/src/main/scala/akka/io/AsyncDnsResolver.scala#L112